server/src/main/java/org/elasticsearch/cluster/metadata/AliasAndIndexLookupBuilder.java

```diff
@@ -0,0 +1,145 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cluster.metadata;
+
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+
+import java.util.Collections;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * Helper class to help build the <code>aliasAndIndexLookup</code> {@link SortedMap}
+ * that is used by several services to retrieve information relating to the relationship
+ * between aliases and indices.
+ *
+ * The {@link SortedMap} holds key-value pairs where the key is either an index name, or
+ * an alias name. Index keys have values of {@link AliasOrIndex.Index}, while alias keys
+ * have values of {@link AliasOrIndex.Alias}. Both {@link AliasOrIndex} types hold a
+ * list of indices that they point to.
+ */
+public class AliasAndIndexLookupBuilder {
+
+    private final SortedMap<String, AliasOrIndex> map;
+
+    public AliasAndIndexLookupBuilder() {
+        this.map = new TreeMap<>();
+    }
+
+    public AliasAndIndexLookupBuilder(SortedMap<String, AliasOrIndex> aliasAndIndexLookup) {
+        this.map = new TreeMap<>(aliasAndIndexLookup);
+    }
+
+    public AliasOrIndex get(String aliasOrIndex) {
+        return map.get(aliasOrIndex);
+    }
+
+    public boolean containsKey(String aliasOrIndex) {
+        return map.containsKey(aliasOrIndex);
+    }
+
+    /**
+     * This method adds an entry for the index represented by <code>indexMetaData</code>, as well
+     * as an entry for each alias that points to it.
+     *
+     * @param indexMetaData the index to add
+     * @throws IllegalStateException if an alias with the same name as an existing index is to be added.
+     */
+    public void addIndex(IndexMetaData indexMetaData) {
+        map.put(indexMetaData.getIndex().getName(), new AliasOrIndex.Index(indexMetaData));
+        for (ObjectObjectCursor<String, AliasMetaData> aliasCursor : indexMetaData.getAliases()) {
+            AliasMetaData aliasMetaData = aliasCursor.value;
+            map.compute(aliasMetaData.getAlias(), (aliasName, alias) -> {
+                if (alias == null) {
+                    return new AliasOrIndex.Alias(aliasMetaData, indexMetaData);
+                } else if (alias instanceof AliasOrIndex.Alias) {
+                    ((AliasOrIndex.Alias) alias).addIndex(indexMetaData);
+                    return alias;
+                } else {
+                    String indexName = ((AliasOrIndex.Index) alias).getIndex().getIndex().getName();
+                    throw new IllegalStateException("index and alias names need to be unique, but the following duplicate was found ["
```

Comment:
I understand I ported this over from `MetaData.Builder#build` where `IllegalStateException` makes more sense. I believe this should be `IllegalAliasNameException`, will update